File: Example/PHP/demo/codeigniter4/Plugins/NextSpanPlugin.php
<?php
#-------------------------------------------------------------------------------
# Copyright 2019 NAVER Corp
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License.  You may obtain a copy
# of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
# License for the specific language governing permissions and limitations under
# the License.
#-------------------------------------------------------------------------------


namespace Plugins;
use Plugins\Candy;


///@hook:CodeIgniter\HTTP\CURLRequest::send
class NextSpanPlugin extends Candy
{
    ///@hook:CodeIgniter\HTTP\CURLRequest::request
    public function onBefore(){
        if(strpos($this->apId, "request"))
        {
            //todo append the pinpoint header
            return ;
        }

        if(strpos($this->apId, "send"))
        {
            pinpoint_add_clue(DESTINATION,parse_url($this->args[1],PHP_URL_PATH));
            pinpoint_add_clue(SERVER_TYPE,PINPOINT_PHP_REMOTE);
            pinpoint_add_clue(NEXT_SPAN_ID,PerRequestPlugins::instance()->getCurNextSpanId());
            pinpoint_add_clues(HTTP_URL,parse_url($this->args[1],PHP_URL_PATH));
        }

    }

    public function onEnd(&$ret)
    {
        if(strpos($this->apId, "send"))
        {
            pinpoint_add_clues(HTTP_STATUS_CODE, (string)$ret->getStatusCode());
        }
    }

    public function onException($e){
        pinpoint_add_clue("EXP",$e->getMessage());
    }

}

File: plugins/PY/pinpoint/plugins/requests/NextSpanPlugin.py
# ------------------------------------------------------------------------------
#  Copyright  2020. NAVER Corp.
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
# ------------------------------------------------------------------------------


from ..common import *
from .PinpointDefine import *
import pinpointPy
from  urllib.parse import urlparse


class NextSpanPlugin(Candy):

    def __init__(self,class_name,module_name):
        super().__init__(class_name,module_name)
        self.nsid = ''
        self.url =''

    def onBefore(self,*args, **kwargs):
        args, kwargs = super().onBefore(*args, **kwargs)
        self.url = args[0]
        generatePinpointHeader(self.url, kwargs['headers'])
        ###############################################################
        pinpointPy.add_clue(FuncName,self.getFuncUniqueName())
        pinpointPy.add_clue(ServerType,PYTHON_METHOD_CALL)
        pinpointPy.add_clues(PY_ARGS, self.url)
        ###############################################################
        return args, kwargs

    def onEnd(self,ret):
        ###############################################################
        pinpointPy.add_clue("dst", urlparse(self.url)['netloc'])
        pinpointPy.add_clue("stp", PYTHON_REMOTE_METHOD)
        pinpointPy.add_clue('nsid', pinpointPy.get_context_key('nsid'))
        pinpointPy.add_clues(HTTP_URL, self.url)
        pinpointPy.add_clues(HTTP_STATUS_CODE, str(ret.status_code))
        pinpointPy.add_clues(PY_RETURN, str(ret))
        ###############################################################
        super().onEnd(ret)
        return ret

    def onException(self, e):
        pinpointPy.add_clue('EXP',str(e))

    def get_arg(self, *args, **kwargs):
        args_tmp = {}
        j = 0

        for i in args:
            args_tmp["arg["+str(j)+"]"] = (str(i))
            j += 1

        for k in kwargs:
            args_tmp[k] = kwargs[k]

        return str(args_tmp)
